bn256/pairing.go

```diff
@@ -122,137 +121,83 @@ func FinalExponentiation(z *GT, _z ...*GT) GT {
 	return result
 }
 
-var lineEvalPool = sync.Pool{
-	New: func() interface{} {
-		return new([88]lineEvaluation)
-	},
-}
-
 // MillerLoop Miller loop
 func MillerLoop(P []G1Affine, Q []G2Affine) (GT, error) {
-	nP := len(P)
-	if nP == 0 || nP != len(Q) {
+	n := len(P)
+	if n == 0 || n != len(Q) {
 		return GT{}, errors.New("invalid inputs sizes")
 	}
 
-	var (
-		ch          = make([]chan struct{}, 0, nP)
-		evaluations = make([]*[88]lineEvaluation, 0, nP)
-		Paff        = make([]G1Affine, nP)
-	)
+	// filter infinity points
+	p := make([]G1Affine, 0, n)
+	q := make([]G2Affine, 0, n)
 
-	var countInf = 0
-	for k := 0; k < nP; k++ {
+	for k := 0; k < n; k++ {
 		if P[k].IsInfinity() || Q[k].IsInfinity() {
-			countInf++
 			continue
 		}
-
-		ch = append(ch, make(chan struct{}, 10))
-		evaluations = append(evaluations, lineEvalPool.Get().(*[88]lineEvaluation))
-
-		Paff[k-countInf].Set(&P[k])
-		go preCompute(evaluations[k-countInf], &Q[k], ch[k-countInf])
+		p = append(p, P[k])
+		q = append(q, Q[k])
+	}
+	n = len(p)
+
+	// projective points for Q
+	qProj := make([]g2Proj, n)
+	qNeg := make([]G2Affine, n)
+	for k := 0; k < n; k++ {
+		qProj[k].FromAffine(&q[k])
+		qNeg[k].Neg(&q[k])
 	}
-
-	nP = nP - countInf
 
 	var result GT
 	result.SetOne()
 
-	j := 0
-	for i := len(loopCounter) - 2; i >= 0; i-- {
+	var l lineEvaluation
 
+	for i := len(loopCounter) - 2; i >= 0; i-- {
 		result.Square(&result)
-		for k := 0; k < nP; k++ {
-			<-ch[k]
-			lineEval(&result, &evaluations[k][j], &Paff[k])
-		}
-		j++
 
-		if loopCounter[i] != 0 {
-			for k := 0; k < nP; k++ {
-				<-ch[k]
-				lineEval(&result, &evaluations[k][j], &Paff[k])
+		for k := 0; k < n; k++ {
+			qProj[k].DoubleStep(&l)
+			lineEval(&result, &l, &p[k])
+
+			if loopCounter[i] == 1 {
+				qProj[k].AddMixedStep(&l, &q[k])
+				lineEval(&result, &l, &p[k])
+			} else if loopCounter[i] == -1 {
+				qProj[k].AddMixedStep(&l, &qNeg[k])
+				lineEval(&result, &l, &p[k])
 			}
-			j++
 		}
 	}
 
+	var Q1, Q2 G2Affine
 	// cf https://eprint.iacr.org/2010/354.pdf for instance for optimal Ate Pairing
-	for k := 0; k < nP; k++ {
-		<-ch[k]
-		lineEval(&result, &evaluations[k][j], &Paff[k])
-	}
-	j++
-	for k := 0; k < nP; k++ {
-		<-ch[k]
-		lineEval(&result, &evaluations[k][j], &Paff[k])
+	for k := 0; k < n; k++ {
+		//Q1 = Frob(Q)
+		Q1.X.Conjugate(&q[k].X).MulByNonResidue1Power2(&Q1.X)
+		Q1.Y.Conjugate(&q[k].Y).MulByNonResidue1Power3(&Q1.Y)
+
+		// Q2 = -Frob2(Q)
+		Q2.X.MulByNonResidue2Power2(&q[k].X)
+		Q2.Y.MulByNonResidue2Power3(&q[k].Y).Neg(&Q2.Y)
+
+		qProj[k].AddMixedStep(&l, &Q1)
+		lineEval(&result, &l, &p[k])
+		qProj[k].AddMixedStep(&l, &Q2)
+		lineEval(&result, &l, &p[k])
 	}
 
-	// release objects into the pool
-	go func() {
-		for i := 0; i < len(evaluations); i++ {
-			lineEvalPool.Put(evaluations[i])
-		}
-	}()
-
 	return result, nil
 }
 
-func lineEval(z *GT, l *lineEvaluation, P *G1Affine) *GT {
+func lineEval(result *GT, l *lineEvaluation, P *G1Affine) *GT {
```

Comment:
We can maybe inline `lineEval()` as in `bls381/`, which makes the Miller Loop on a par with `develop` branch (w/ goroutines).

Reply:
resolved in 5ab1d22f07799ac023679d42093b265e5c1e58cb.